### PR TITLE
[improvement](tcmalloc) add moderate mode and avoid oom with a lot of…

### PR DESCRIPTION
**Pull request by dataroaring**
… cache

ReleaseToSystem aggressively when there are little free memory.

# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/common/daemon.cpp [hunk 1]
```diff
@@ -74,28 +74,49 @@ void Daemon::tcmalloc_gc_thread() {
...
+
+            if (MemInfo::mem_limit() > alloc_size) {
+                max_free_size = MemInfo::mem_limit() - alloc_size;
+                interval_seconds = 2;
+            } else {
+                interval_seconds = 1;
+            }
+
+            max_free_size = std::min(used_size * 20 / 100, max_free_size);
+            size_t free_size = alloc_size - used_size - max_free_size;
```
Comment (xinyiZzz) [inline be/src/common/daemon.cpp:116]: It seems difficult to understand using `used_size` to calculate `max_free_size`, and then to calculate `free_size`.
Directly free the percentage of `MemInfo::allocator_cache_mem()`. Seems to be better understood.

Comment (xinyiZzz) [reply be/src/common/daemon.cpp:116]: E.g,
1) When `alloc_size` < `mem_limit` * 0.8, release `MemInfo::allocator_cache_mem()` * 0.2 each time
2) When `alloc_size` > `mem_limit` * 0.8, release `MemInfo::allocator_cache_mem()` * 0.8 each time

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.78 seconds
 load time: 436 seconds
 storage size: 17150999303 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221117135918_clickbench_pr_48127.html

**Review by k-i-d-d — DISMISSED** [review]
LGTM, good code

**Review by xinyiZzz — DISMISSED** [review]
LGTM

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Pull request merged**